OpenRA.Mods.Common/Traits/Render/ProductionBar.cs

```diff
@@ -17,75 +17,63 @@
 namespace OpenRA.Mods.Common.Traits.Render
 {
 	[Desc("Visualizes the remaining build time of actor produced here.")]
-	class ProductionBarInfo : ITraitInfo, Requires<ProductionInfo>
+	class ProductionBarInfo : ConditionalTraitInfo, Requires<ProductionInfo>
 	{
+		[FieldLoader.Require]
 		[Desc("Production queue type, for actors with multiple queues.")]
 		public readonly string ProductionType = null;
 
 		public readonly Color Color = Color.SkyBlue;
 
-		public object Create(ActorInitializer init) { return new ProductionBar(init.Self, this); }
+		public override object Create(ActorInitializer init) { return new ProductionBar(init.Self, this); }
 	}
 
-	class ProductionBar : ISelectionBar, ITick, INotifyCreated, INotifyOwnerChanged
+	class ProductionBar : ConditionalTrait<ProductionBarInfo>, ISelectionBar, ITick, INotifyCreated
 	{
-		readonly ProductionBarInfo info;
 		readonly Actor self;
 		ProductionQueue queue;
 		float value;
 
 		public ProductionBar(Actor self, ProductionBarInfo info)
+			: base(info)
 		{
 			this.self = self;
-			this.info = info;
 		}
 
-		void FindQueue()
+		void INotifyCreated.Created(Actor self)
 		{
-			var type = info.ProductionType ?? self.Info.TraitInfo<ProductionInfo>().Produces.First();
-
 			// Per-actor queue
 			// Note: this includes disabled queues, as each bar must bind to exactly one queue.
 			queue = self.TraitsImplementing<ProductionQueue>()
-				.FirstOrDefault(q => type == null || type == q.Info.Type);
+				.FirstOrDefault(q => Info.ProductionType == q.Info.Type);
 
 			if (queue == null)
 			{
 				// No queues available - check for classic production queues
 				queue = self.Owner.PlayerActor.TraitsImplementing<ProductionQueue>()
-					.FirstOrDefault(q => type == null || type == q.Info.Type);
+					.FirstOrDefault(q => Info.ProductionType == q.Info.Type);
 			}
-
-			if (queue == null)
```

Comment:
You should also keep this check around, incase the modder typo.

Reply:
We (should) have lint tests for that!